doc/SAI_failure_handling/SAI_failure_handling.md

```diff
@@ -0,0 +1,100 @@
+# 1. Scope
+This document describes the high-level design for orchagent in handling SAI failures
+
+
+# 2. Failure Handling Framework
+### 2.1 Requirements for failure handling functions in orchagent
+1. Allow different handling for Create/Set/Remove operations.
+1. Allow each Orch to have its specific handling logic.
+1. Adapt handling logic based on SAI API type and SAI status.
+1. Escalate the failure to upper layers when the failure cannot be handled in orchagent.
+
+### 2.2 An overview of the failure handling framework
+An illustrative figure of the failure handling framework is shown below.
+The orchagent generates SAI calls according to the information in APPL_DB given by upper layers.
+In the case of SAI failures, the orchagent gets the failure status via the feedback mechanism in synchronous mode.
+Based on the failure information, the failure handling functions in orchagent make the first attempt to address the failure.
+An ERROR_DB is also introduced to support escalation to upper layers.
+In the scenario where orchagent is unable to resolve the problem, the failure handling functions would escalate the failure to upper layers by pushing the failure into the ERROR_DB.
+<img src="Framework.png">
+
+### 2.3 Failure handling functions in Orchagent
+#### 2.3.1 Failure handling functions
+To support a failure handling logic in general while also allow each orch to have its specific logic, we include the following virtual functions in Orch
+1. `virtual task_process_status handleSaiCreateStatus(sai_api_t api, sai_status_t status, void *context = nullptr)`
+2. `virtual task_process_status handleSaiSetStatus(sai_api_t api, sai_status_t status, void *context = nullptr)`
+3. `virtual task_process_status handleSaiRemoveStatus(sai_api_t api, sai_status_t status, void *context = nullptr)`
+
+The three functions handle SAI failures in create, set, and remove operations, respectively.
+With the type of SAI API and SAI status as an input, the function could handle the failure according to the two pieces ofinformation.
+
+In the scenario where a specific logic is required in one of the Orchs, this design allows the Orch to inherit the function and include the specific login in the inherited function.
+
+The function also allows an optional input `context`, which allows passing context (e.g., object entry, attribute, etc.) into the function so that it could escalate the information to the ERROR_DB and upper layers.
+
+#### 2.3.2 Possible execution results
+1. Return `task_success` --  No crash, no retry, handled successfully. 
+
+The failure handling function should return `task_success` when the failure is properly handled without the need for another attempt (e.g., the SAI status is `SAI_STATUS_ITEM_NOT_FOUND` in remove operation).
+
+2. Return `task_failed` -- No crash, no retry, not handled successfully. 
```

Comment:
>task_failed [](start = 11, length = 11)

There are seveal causes for task_failed
1. user input is invalid, such as wrong ACL, conflicting IP addresses
2. hardware permanent error
3. internal logic error, but not critical, and we still want to keep dataplane working
3. etc.

Reply:
You can mentioned 'no exit' or 'keep running'

---
In reply to: [600051934](https://github.com/Azure/SONiC/pull/762#discussion_r600051934) [](ancestors = 600051934)

Reply:
Added the scenarios of failure and clarified the keep running behavior.